store/mockstore/mocktikv/cop_handler_dag.go

```diff
@@ -585,19 +594,110 @@ func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, *cop
 		OutputCounts: counts,
 	}
-	if len(warnings) > 0 {
-		selResp.Warnings = make([]*tipb.Error, 0, len(warnings))
-		for i := range warnings {
-			selResp.Warnings = append(selResp.Warnings, toPBError(warnings[i].Err))
+	for i := range warnings {
+		selResp.Warnings = append(selResp.Warnings, toPBError(warnings[i].Err))
+	}
+	return selResp
+}
+
+func (h *rpcHandler) supplementSelectResponse(selResp *tipb.SelectResponse, dagReq *tipb.DAGRequest, dagCtx *dagContext, rows [][][]byte) error {
+	colTypes := h.constructRespSchema(dagCtx)
+	loc := dagCtx.evalCtx.sc.TimeZone
+
+	switch dagReq.EncodeType {
+	case tipb.EncodeType_TypeDefault:
+		h.encodeDefault(selResp, rows, dagReq.OutputOffsets)
+	case tipb.EncodeType_TypeArrow:
+		err := h.encodeArrow(selResp, rows, colTypes, dagReq.OutputOffsets, loc)
+		if err != nil {
+			return errors.Trace(err)
 		}
 	}
+	return nil
+}
+
+func (h *rpcHandler) constructRespSchema(dagCtx *dagContext) []*types.FieldType {
+	root := dagCtx.dagReq.Executors[len(dagCtx.dagReq.Executors)-1]
+	if root.Aggregation != nil {
+		hashAgg := root.Aggregation
+		schema := make([]*types.FieldType, 0, len(hashAgg.AggFunc)+len(hashAgg.GroupBy))
+		for i := range hashAgg.AggFunc {
+			if hashAgg.AggFunc[i].Tp == tipb.ExprType_Avg {
+				schema = append(schema, types.NewFieldType(mysql.TypeLonglong))
+			}
+			schema = append(schema, expression.PbTypeToFieldType(hashAgg.AggFunc[i].FieldType))
+		}
+		for i := range hashAgg.GroupBy {
+			schema = append(schema, expression.PbTypeToFieldType(hashAgg.GroupBy[i].FieldType))
+		}
+		return schema
+	} else if root.StreamAgg != nil {
+		streamAgg := root.StreamAgg
+		schema := make([]*types.FieldType, 0, len(streamAgg.AggFunc)+len(streamAgg.GroupBy))
+		for i := range streamAgg.AggFunc {
+			if streamAgg.AggFunc[i].Tp == tipb.ExprType_Avg {
+				schema = append(schema, types.NewFieldType(mysql.TypeLonglong))
+			}
+			schema = append(schema, expression.PbTypeToFieldType(streamAgg.AggFunc[i].FieldType))
+		}
+		for i := range streamAgg.GroupBy {
+			schema = append(schema, expression.PbTypeToFieldType(streamAgg.GroupBy[i].FieldType))
+		}
+		return schema
+	}
+	return dagCtx.evalCtx.fieldTps
+}
+
+func (h *rpcHandler) encodeDefault(selResp *tipb.SelectResponse, rows [][][]byte, colOrdinal []uint32) {
+	var chunks []tipb.Chunk
+	for i := range rows {
+		requestedRow := dummySlice
+		for _, ordinal := range colOrdinal {
+			requestedRow = append(requestedRow, rows[i][ordinal]...)
+		}
+		chunks = appendRow(chunks, requestedRow, i)
+	}
+	selResp.Chunks = chunks
+}
+
+func (h *rpcHandler) encodeArrow(selResp *tipb.SelectResponse, rows [][][]byte, colTypes []*types.FieldType, colOrdinal []uint32, loc *time.Location) error {
+	rowBatchData := make([]byte, 0, 1024)
+	respColTypes := make([]*types.FieldType, 0, len(colOrdinal))
+	for _, ordinal := range colOrdinal {
+		respColTypes = append(respColTypes, colTypes[ordinal])
+	}
+
+	chk := chunk.NewChunkWithCapacity(respColTypes, 32)
```

Comment:
where does the `32` come from? Should it be equal to `rowsPerChunk`?

Reply:
Uh.. Yes, it should better be equal to `rowsPerChunk`. I'll refactor this. BTW, `32` does not affect the correctness.